sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -164,6 +164,10 @@ object Cast extends QueryErrorsBase {
 
     case (udt1: UserDefinedType[_], udt2: UserDefinedType[_]) if udt2.acceptsType(udt1) => true
 
+    // Casting from GEOGRAPHY to GEOMETRY with the same SRID is allowed.
+    case (geog: GeographyType, geom: GeometryType) if geog.srid == geom.srid =>
```

Comment:
We have a special srid that means ANY. Shall we allow casting from certain srid to ANY?